clar2wasm/src/tools.rs

```diff
@@ -178,6 +181,108 @@ pub fn evaluate_at(
+        Value::Sequence(SequenceData::List(list_data)) => {
+            let mut processed_data = Vec::new();
+
+            for val in &list_data.data {
+                processed_data.push(unicode_to_byte_sequence(val.clone()));
+            }
+
+            Value::Sequence(SequenceData::List(ListData {
+                data: processed_data,
+                type_signature: list_data.type_signature.clone(),
+            }))
+        }
+        Value::Response(response_data) => Value::Response(ResponseData {
+            committed: response_data.committed,
+            data: Box::new(unicode_to_byte_sequence(*(response_data.data).clone())),
+        }),
+        Value::Optional(optional_data) => {
+            if let Some(data) = &optional_data.data {
+                Value::Optional(OptionalData {
+                    data: Some(Box::new(unicode_to_byte_sequence(*data.clone()))),
+                })
+            } else {
+                Value::Optional(optional_data.clone())
+            }
+        }
+        Value::Tuple(tuple_data) => {
+            let mut btree_map = BTreeMap::new();
+            for (name, value) in tuple_data.data_map.iter() {
+                btree_map.insert(name.clone(), unicode_to_byte_sequence(value.clone()));
+            }
+            Value::Tuple(TupleData {
+                data_map: btree_map,
+                type_signature: tuple_data.type_signature.clone(),
+            })
+        }
+        Value::Sequence(SequenceData::String(CharType::UTF8(u))) => {
+            println!("Input Data :{:?}", u.data);
+
+            let scalar_bytes = u.data.clone();
+
+            let mut utf8_bytes: Vec<Vec<u8>> = Vec::new();
+
+            for bytes in scalar_bytes {
+                // Reconstruct the scalar value assuming big-endian byte order
+                let mut scalar_value: u32 = 0;
+                let num_bytes = bytes.len();
+                for (i, &byte) in bytes.iter().enumerate() {
+                    scalar_value |= (byte as u32) << (8 * (num_bytes - 1 - i));
+                }
+                println!("Scalar value: {scalar_value}");
+                // Convert the scalar value to a char if it's a valid Unicode scalar value
+                if let Some(character) = std::char::from_u32(scalar_value) {
+                    if character.is_ascii() {
+                        utf8_bytes.push(vec![character as u8]);
+                    } else {
+                        let mut buf = [0; 4];
+                        let encoded_str = character.encode_utf8(&mut buf);
+                        utf8_bytes.push(encoded_str.as_bytes().to_vec());
+                    }
+                }
+            }
+
+            println!("Output UTF-8 Data: {:?}", utf8_bytes);
+
+            Value::Sequence(SequenceData::String(CharType::UTF8(UTF8Data {
+                data: utf8_bytes,
+            })))
+        }
+        _ => input,
+    }
+}
+
+pub fn unicode_scalars_from_string(str: String) -> Value {
```

Comment:
The PR changes code which previously used `Value::string_utf8_from_bytes` to instead use this function, but it seems like the implementation is different. Why not use `Value::string_utf8_from_bytes`?

Reply:
Yes. Now I'm expecting my utf-8 string to have Unicode bytes in crosscheck method. This method is to convert string to those Unicode bytes.

Reply:
What do you mean by "Unicode bytes"? Utf8 encodes unicode, the `Value::string_utf8_from_bytes` function encodes unicode into the `Value::Sequence(SequenceData::String(CharType::UTF8))` type, clarity-wasm code encodes unicode into 4-byte scalar values, etc..

Reply:
> What do you mean by "Unicode bytes"?

I meant unicode scalars (in Vec<Vec<u8>> form).

> Utf8 encodes unicode, the `Value::string_utf8_from_bytes` function encodes unicode into the `Value::Sequence(SequenceData::String(CharType::UTF8))`

I got that. Simple question, How should I generate Unicode then?

Currently I'm doing something like this

```rust
        let mut data = vec![];
        for char in str.chars() {
            let scalar_value = char as u32; // Get the Unicode scalar value of the character
            let mut scalar_bytes = scalar_value.to_be_bytes().to_vec(); // Convert scalar value to big-endian bytes
            scalar_bytes.retain(|&x| x != 0); // Remove trailing zeros
            data.push(scalar_bytes);
        }
```

Reply:
Your still using phrases that don't make sense like "should I generate Unicode?". Unicode is just a standard of code points..

Are you asking how to create a Clarity utf8-string literal..? If so, just escape the non-ascii code points e.g.
```rs
fn escape_non_ascii(input: &str) -> String {
    input
        .chars()
        .map(|c| {
            if c.is_ascii() {
                c.to_string()
            } else {
                format!("\\u{{{:X}}}", c as u32)
            }
        })
        .collect()
}

fn main() {
    let input = "Hello, 世界!";
    let escaped = escape_non_ascii(input);
    println!("{}", escaped); // Hello, \u{4E16}\u{754C}!
}
```